metrics/engine/ingester.go

```diff
@@ -92,6 +97,60 @@ func (oi *outputIngester) flushMetrics() {
 				oi.metricsEngine.markObserved(sm.Metric)
 				sm.Metric.Sink.Add(sample)
 			}
+
+			oi.cardinality.Add(sample.TimeSeries)
 		}
 	}
+
+	if oi.cardinality.LimitHit() {
+		// TODO: suggest using the Metadata API as an alternative, once it's
+		// available (e.g. move high-cardinality tags as Metadata)
+		// https://github.com/grafana/k6/issues/2766
+
+		oi.logger.Warnf(
+			"The test has generated metrics with %d unique time series, "+
+				"which is higher than the suggested limit of %d "+
+				"and could cause high memory usage. "+
+				"Consider not using high-cardinality values like unique IDs as metric tags "+
+				"or, if you need them in the URL, use the name metric tag or URL grouping. "+
+				"See https://k6.io/docs/using-k6/tags-and-groups for details.", oi.cardinality.Count(), timeSeriesFirstLimit)
+	}
+}
+
+type cardinalityControl struct {
+	seen            map[metrics.TimeSeries]struct{}
+	timeSeriesLimit int
+}
+
+func newCardinalityControl() *cardinalityControl {
+	return &cardinalityControl{
+		timeSeriesLimit: timeSeriesFirstLimit,
+		seen:            make(map[metrics.TimeSeries]struct{}),
+	}
+}
+
+// Add adds the passed time series to the list of seen items.
+func (cc *cardinalityControl) Add(ts metrics.TimeSeries) {
+	if _, ok := cc.seen[ts]; ok {
+		return
+	}
+	cc.seen[ts] = struct{}{}
+}
+
+// LimitHit checks if the cardinality limit has been hit.
+func (cc *cardinalityControl) LimitHit() bool {
+	if len(cc.seen) <= cc.timeSeriesLimit {
+		return false
+	}
+
+	// we don't care about overflow
+	// the process should be already OOM
+	// if the number of generated time series goes higher than N-hundred-million(s).
+	cc.timeSeriesLimit *= 2
```

Comment:
Is the idea that we want them to get the message multiple times as they continue to make more time series?

I am not really against that, just in the case that we forgo this, we can actually stop counting once the limit is hit and even throw away the `seen` map. 

Not really certain if users will be better off seeing this on 100k, 200k, 400k, 800k, 1.6m and so on time series.

Reply:
:thinking: yeah, maybe we can show that warning 5 times and then stop tracking time series :thinking: not that such an optimization will help very much at 3.2million time series, but still :sweat_smile:

Reply:
I'd argue that it would be useful to see it a few times, since the time between warnings would give information about how quickly the number of time series grows, which is impossible to see or guess otherwise with local k6 runs, and not always easy even with an external output.

Reply:
I think it is also helpful for us, in the case of reports, to see what it is the common order of magnitude. We should act differently if they often hit around millions instead of thousands.